### DISCUSSION TITLE
Is there a way to install the virtual store but not create node_module symlinks?

### DISCUSSION THREAD

**Discussion by jfirebaugh** [pnpm/pnpm#7294] [Q&A]
For CI optimization purposes, I would like to cache the virtual store. Yes, I know the content-addressable store is the recommended thing to cache. But in practice, in large monorepos, even cloning or hardlinking from CAS to virtual store can take upwards of 20 seconds or more if the virtual store is empty. In other words, caching the virtual store produces faster builds than caching the CAS. Symlinking a workspace package's node_modules to the virtual store is negligible in comparison and can be uncached.

Is there a way to install the virtual store but _not_ create symlinked node_modules for any workspace packages?

The reason I want to avoid creating any node_module symlinks is that I would like to be able to create a fresh cache artifact in any CI job which detects that the checksum of pnpm-lock.yaml has changed. The job that ends up saving the cache artifact typically wants to do `pnpm install --filter ./package/foo` to avoid time spent symlinking anything but what's needed for that job. (Although unfortunately `pnpm install --filter` is currently [not working](https://github.com/pnpm/pnpm/issues/7242).) The desired caching algorithm is:

1. If a cache artifact for the key `pnpm-virtual-store-$(checksum pnpm-lock.yaml)` exists, restore it.
2. Otherwise, run `pnpm install --only-virtual-store` (hypothetical option that creates a fully populated `node_modules/.pnpm` but nothing else). Store `node_modules/.pnpm` to the key `pnpm-virtual-store-$(checksum pnpm-lock.yaml)`.
3. Run `pnpm install --filter <desired filter>`. This should be fast as the only thing it needs to do is generate symlinks for a subset of workspace packages.

**Comment by zkochan** [comment]
Can you validate at all if this will work? You can validate it with the symlinks present too.

**Reply by jfirebaugh** [reply]
Yes, it will work, I've tried it with the symlinks present.
